torch/csrc/autograd/python_variable.cpp

```diff
@@ -2036,4 +2038,21 @@ c10::Device concrete_device_fn(const c10::impl::PyInterpreter*, const c10::Tenso
   return toDevice(out.ptr());
 }
 
+c10::IntArrayRef concrete_strides_fn(const c10::impl::PyInterpreter*, const c10::TensorImpl* self) {
+  pybind11::gil_scoped_acquire gil;
+  at::impl::MaybeSetTLSOnEntryGuard guard;
+
+  auto out = torchDispatchFromTensorImpl(
+      self,
+      "stride",
+      py::module::import("torch")
+          .attr("ops")
+          .attr("aten")
+          .attr("stride")
+          .ptr(),
+      "torch.ops.aten");
+
+  return c10::IntArrayRef(THPUtils_unpackLongs(out.ptr()));
```

Comment:
So this is pretty interesting - the TensorImpl API's for sizes/strides return non-owning `IntArrayRef` references. Which works fine for normal C++ TensorImpl, because the lifetime of the metadata is tied to the lifetime of the tensor.

With `__torch_dispatch__` though, the thing being returned is a fresh pyobject who's lifetime is **not** tied to the tensor it came from (at least, that's my nerd snipe / guess). We're stuck though with the fact that `TensorImpl::strides()` is supposed to return a non-owning `IntArrayRef`.

One question that I have is - why do we want to be able to override strides from python? Do really want the internal strides in C++ to diverge from what the user sees when they call `.strides()`?

If not, then hot take: when you override sizes/strides in __torch_dispatch__ to return something else, morally, you're telling what the **real** sizes/strides of that tensor are. So we could copy the returned strides directly onto the TensorImpl. Does that sound reasonable? One immediate downside is that calling `.strides()` on a tensor subclass is no longer thread safe, since we're mutating the tensor metadata (we could use locks and fix that though).

cc @albanD

Reply:
oh, just saw the new comment :). I think I vote for "figure out a way to ensure the strides are updated on the TensorImpl"

Reply:
@ezyang says he doesn't like this idea, I'll leave it to him to elaborate further (although I imagine it's because you have a complicated split of source of truth between the Python tensor and the value stored in TensorImpl, which basically creates a "cache-invalidation-but-harder" problem).

Another option is that the user is required to provide an object from which we can construct an IntArrayRef and is responsible for making sure the lifetime is bound to the lifetime of the tensor. You could accomplish this through ctypes; you can construct an array type like:
```
from ctypes import *
SizeType = c_long * ndim
foo = SizeType(dim1, dim2)
```
The memory layout of this is the same as `long foo[ndim]`, so we can construct an IntArrayRef out of it.

If you attach that to your PythonTensor instance and return it from `__torch_dispatch__`, you should be good, right?